ethcore/private-tx/src/lib.rs

```diff
@@ -261,48 +251,125 @@ impl Provider where {
+					if let Action::Call(contract) = tx_action {
 						// TODO [ToDr] Usage of BlockId::Latest
-						let contract_nonce = self.get_contract_nonce(&contract, BlockId::Latest)?;
-						let private_state = self.execute_private_transaction(BlockId::Latest, transaction.signed())?;
+						let contract_nonce = self.get_contract_nonce(&contract, BlockId::Latest);
+						if let Err(e) = contract_nonce {
+							bail!("Cannot retrieve contract nonce: {:?}", e);
+						}
+						let contract_nonce = contract_nonce.expect("Error was checked before");
+						let private_state = self.execute_private_transaction(BlockId::Latest, &transaction.transaction);
+						if let Err(e) = private_state {
+							bail!("Cannot retrieve private state: {:?}", e);
+						}
+						let private_state = private_state.expect("Error was checked before");
 						let private_state_hash = self.calculate_state_hash(&private_state, contract_nonce);
 						trace!("Hashed effective private state for validator: {:?}", private_state_hash);
-						let password = find_account_password(&self.passwords, &*self.accounts, &account);
-						let signed_state = self.accounts.sign(account, password, private_state_hash)?;
-						let signed_private_transaction = SignedPrivateTransaction::new(desc.private_hash, signed_state, None);
+						let password = find_account_password(&self.passwords, &*self.accounts, &validator_account);
+						let signed_state = self.accounts.sign(validator_account, password, private_state_hash);
+						if let Err(e) = signed_state {
+							bail!("Cannot sign the state: {:?}", e);
+						}
+						let signed_state = signed_state.expect("Error was checked before");
+						let signed_private_transaction = SignedPrivateTransaction::new(private_hash, signed_state, None);
 						trace!("Sending signature for private transaction: {:?}", signed_private_transaction);
-						self.broadcast_signed_private_transaction(signed_private_transaction.rlp_bytes().into_vec());
+						self.broadcast_signed_private_transaction(signed_private_transaction.hash(), signed_private_transaction.rlp_bytes().into_vec());
 					} else {
-						warn!("Incorrect type of action for the transaction");
+						bail!("Incorrect type of action for the transaction");
 					}
-				},
-				Err(e) => {
-					warn!("Cannot retrieve descriptor for transaction with error {:?}", e);
 				}
 			}
-			verification_queue.remove_private_transaction(&transaction_hash);
+			return Ok(());
+		};
+		let ready_transactions = self.transactions_for_verification.drain(self.pool_client(&nonce_cache));
+		for transaction in ready_transactions {
+			if let Err(e) = process_transaction(&transaction) {
+				warn!("Error: {:?}", e);
+			}
 		}
 		Ok(())
 	}
 
+	/// Add signed private transaction into the store
+	/// Creates corresponding public transaction if last required signature collected and sends it to the chain
+	pub fn process_signature(&self, signed_tx: SignedPrivateTransaction) -> Result<(), Error> {
```

Comment:
take ```signed_tx``` by reference instead and get rid of a ```clone()```: https://github.com/paritytech/parity-ethereum/blob/2778d4e99bb25c8e2df3f15617e57b16156fbc10/ethcore/private-tx/src/lib.rs#L678